### PR TITLE
Fix visual mode bugs#1304to#1308

### LINKED ISSUE #1306: Visual Mode Issues: S command
### What did you do?

Visually selected this text and pressed S, brackets represent the visual selection:
Line before
hell[man's
lma]o world
Line after

### What did you expect to happen?

In Vim, the selected lines are removed and the editor is placed in insert mode:
Line before

Line after

### What happened instead?

The editor goes into "Surround Insert Mode" then the next character pressed is inserted around the selection, for example s* gives:
Line before
hell*man's
lma*o world
Line after

### Technical details:
* VSCode Version: 1.9.1
* VsCodeVim Version: 0.5.3
* OS: Fedora 25

### PR DESCRIPTION
**Pull request by xlaech**
Fixed the bugs in Visual mode:
- #1304 
- #1305 
- #1307 
- #1308 

Did not fix #1306. See Diskussion.

Please intensively review #1308, since this was the most difficult one and i am still not sure if i have forgotten any edge cases. The tests run thought 👍

### CODE REVIEW

**Comment by xconverge** [comment]
Not ignored ;) he is still on vacation

**Comment by xlaech** [comment]
Haha 👍 
No Problem... won't make a difference if we merge it now or in 2 weeks ;)

File: src/actions/actions.ts [hunk 1]
```diff
@@ -4980,7 +4990,35 @@ class ActionDeleteLineVisualMode extends BaseCommand {
   keys = ["X"];
 
   public async exec(position: Position, vimState: VimState): Promise<VimState> {
-    return await new DeleteOperator(this.multicursorIndex).run(vimState, position.getLineBegin(), position.getLineEnd());
+    if (vimState.currentMode === ModeName.Visual) {
+      return await new DeleteOperator(this.multicursorIndex).run(vimState,
+                                                                vimState.lastVisualSelectionStart.getLineBegin(),
```
Comment (johnfn) [inline src/actions/actions.ts]: Why is it `lastVisualSelection`? Is the current selection not what we want?

`lastSelection` was intended for commands like gv that are triggered after you stopped selecting entirely and did something else, so I just wanna make sure this is correct.

Comment (johnfn) [inline src/actions/actions.ts]: Like ideally you should be able to just access the current selection?

Comment (xlaech) [inline src/actions/actions.ts]: I try it out tomorrow. Since i am not very familiar with the code, i just search with intellisense for possible members/function, i could use 😃

Comment (xlaech) [inline src/actions/actions.ts]: Which properties you want to use?
I can find a current selection member in `vimState`.

Using `TextEditor.getSelection() `does not work.

Comment (johnfn) [inline src/actions/actions.ts]: vimState.cursorStartPosition and vimState.cursorPosition. Sorry I didn't make that more clear and you had to hunt around.

**Comment by xconverge** [comment]
This one seems so close, any chance you could get to the last few things @johnfn mentioned?

**Comment by xlaech** [comment]
Sorry... i worked at another company in the last month and did not have much time. I get to it in 2 weeks :)

**Comment by xconverge** [comment]
Great! Will leave it open then!

**Comment by xlaech** [comment]
I merged the current master branch into mine. But the action.ts file does no longer exist.
Is this due to a change in the projectstructure? Where does the "old" function now reside?

**Comment by xconverge** [comment]
@xlaech it was split between a few files here https://github.com/VSCodeVim/Vim/tree/master/src/actions

I would advise searching for the actions you want with ```"X"``` for example

**Comment by xlaech** [comment]
There you go :) I was very confused by the merge... so i just wrote the whole fix from scratch.

I also came to the conclusion, that the whole put command might need some refactoring in the future. At the moment, I think it is responsible for 2 many things.
I honestly did not make it much better, since i now added the extra case for VisualMode + Linewise register.
Maybe you have some Ideas about that.

Please provide feedback.

File: test/mode/modeVisual.test.ts [hunk 2]
```diff
@@ -810,4 +810,136 @@ suite("Mode Visual", () => {
     });
 
   });
-});
+
+  suite("search works in visual mode", () => {
```
Comment (xconverge) [inline test/mode/modeVisual.test.ts]: why are these 2 tests here? I think they were already merged into master, probably an artifact from your merge? :)

Comment (xlaech) [inline test/mode/modeVisual.test.ts]: I just copied the current file from master in the main repo and rewrote the test. I thought they were added by you guys. I they are not needed i can delete them.

Comment (xconverge) [inline test/mode/modeVisual.test.ts]: But if they are in master already they shouldn't show up in the diff

Comment (Chillee) [inline test/mode/modeVisual.test.ts]: It was just an artifact of an unresolved merge conflict.

**Comment by Chillee** [comment]
So is this good to go now?
